modules/js/perf/perf_imgproc/perf_cvtcolor.js

```diff
@@ -461,60 +475,105 @@ cv.onRuntimeInitialized = () => {
+      log('elapsed time:' +String(event.target.times.elapsed*1000)+' ms');
+      log('mean time:' +String(event.target.stats.mean*1000)+' ms');
+      log('stddev time:' +String(event.target.stats.deviation*1000)+' ms');
+      log(String(event.target));
     })
-    .on('error', function(event) { console.log(`test case ${event.target.name} failed`); })
+    .on('error', function(event) { log(`test case ${event.target.name} failed`); })
     .on('complete', function(event) {
-      console.log(``);
-      console.log(`###################################`);
-      console.log(`Finished testing ${event.currentTarget.length} cases`);
-    })
+      log(`\n ###################################`)
+      log(`Finished testing ${event.currentTarget.length} cases \n`);
+      if (!isNodeJs) {
+        runButton.removeAttribute('disabled');
+        runButton.setAttribute('class', 'btn btn-primary');
+        runButton.innerHTML = 'Run';
+      }
+    });
+  }
+
+
 
-    // run async
-    .run({ 'async': true });
+  // init
+  let cvtFunc = [addCvtModeCase, addCvtModeBayerCase, addCvtMode2Case, addCvtMode3Case];//, addEdgeAwareBayerModeCase];
+  let combinations = [combiCvtMode, combiCvtModeBayer, combiCvtMode2, combiCvtMode3];//, combiEdgeAwareBayer];
+
+  // Flags
+  // set test filter params
+  if (isNodeJs) {
+    let suite = new Benchmark.Suite;
+    totalCaseNum = 0;
+    currentCaseId = 0;
+    const args = process.argv.slice(2);
+    if (/--test_param_filter=\([0-9]+x[0-9]+,[\ ]*\w+\)/g.test(args.toString())) {
+      let params = args.toString().match(/--test_param_filter=\([0-9]+x[0-9]+,[\ ]*\w+\)/g)[0];
+      decodeParams2Case(suite, params);
+    } else {
+      log("no filter or getting invalid params, run all the cases");
+      addCvtModeCase(suite, combiCvtMode);
+      addCvtModeBayerCase(suite, combiCvtModeBayer);
+      addCvtMode2Case(suite, combiCvtMode2);
+      addCvtMode3Case(suite, combiCvtMode3);
+    }
+    setBenchmarkSuite(suite);
+    log(`Running ${totalCaseNum} tests from CvtColor`);
+    suite.run({ 'async': true }); // run the benchmark
+  } else {
+    let suite;
+    runButton.onclick = function()　{
+      suite = new Benchmark.Suite;
+      totalCaseNum = 0;
+      currentCaseId = 0;
+      let paramsContent = paramsElement.value;
+      if (/\([0-9]+x[0-9]+,[\ ]*\w+\)/g.test(paramsContent.toString())) {
+        let params = paramsContent.toString().match(/\([0-9]+x[0-9]+,[\ ]*\w+\)/g)[0];
+        decodeParams2Case(suite, params);
+      } else {
+        log("no filter or getting invalid params, run all the cases");
+        addCvtModeCase(suite, combiCvtMode);
+        addCvtModeBayerCase(suite, combiCvtModeBayer);
+        addCvtMode2Case(suite, combiCvtMode2);
+        addCvtMode3Case(suite, combiCvtMode3);
+      }
+      setBenchmarkSuite(suite);
+      log(`Running ${totalCaseNum} tests from CvtColor`);
+      suite.run({ 'async': true }); // run the benchmark
```

Comment:
I suggest to combine the common code for Node.js and Browser from `suite = new Benchmark.Suite;` to `suite.run({ 'async': true });`. Please avoid the code duplication as much as possible.

Reply:
Yeah. I abstracted the same code into a function named `genBenchmarkCase`. There are some difference between two version so it seems that these two part can't be combined totally.